src/TorchSharp/Tensor/Tensor.cs

```diff
@@ -7402,5 +7402,15 @@ public static Tensor WrappedTensorDisposeScope(Func<Tensor> expr)
             var result = expr();
             return result.MoveToOuterDisposeScope();
         }
+        internal static Tensor InstantiateTensorWithLeakSafeTypeChange(IntPtr handle, ScalarType? dtype)
+        {
+            var tensor = new Tensor(handle);
+            if (dtype.HasValue) {
+                var typed = tensor.to_type(dtype.Value);
+                tensor.Dispose();
```

Comment:
Can you explain why the tensor is orphaned? Why does it not get added to the DisposeScope?

Reply:
What was happening prior is that the `var tensor` is added to the scope if one was present. Then when either the device or dtype branch was hit, a new tensor would be created. The old tensor was never disposed and it's reference never returned, although it is no longer needed. It would remain on the DisposeScope until the scope got disposed. 

If you had no dispose scope, the tensor could last until the application terminates.

Reply:
I see. So, it wasn't really orphaned, then, its lifespan wasn't optimal. In that case, doesn't it have to be detached from any DisposeScope that might have picked it up?

Reply:
Your description is better than mine. Inside the Dispose of Tensor the DisposeScopeManager is notified to update statistics, and the call to OwningDisposeScope.MarkAsDisposed method takes care of removing the object from the DisposeScope that owns it. In the event there was never a DisposeScope, it is still properly disposed.

Reply:
Yes, you're right. I forgot that Tensor.Dispose() had that logic.